crates/wasi/src/preview2/stream.rs

```diff
@@ -151,60 +133,354 @@ pub trait OutputStream: Send + Sync {
+    fn push_output_stream(
+        &mut self,
+        ostream: Box<dyn HostOutputStream>,
+    ) -> Result<u32, TableError> {
         self.push(Box::new(ostream))
     }
-    fn get_output_stream(&self, fd: u32) -> Result<&dyn OutputStream, TableError> {
-        self.get::<Box<dyn OutputStream>>(fd).map(|f| f.as_ref())
+    fn get_output_stream_mut(
+        &mut self,
+        fd: u32,
+    ) -> Result<&mut Box<dyn HostOutputStream>, TableError> {
+        self.get_mut::<Box<dyn HostOutputStream>>(fd)
     }
-    fn get_output_stream_mut(&mut self, fd: u32) -> Result<&mut Box<dyn OutputStream>, TableError> {
-        self.get_mut::<Box<dyn OutputStream>>(fd)
+}
+
+/// Provides a [`HostInputStream`] impl from a [`tokio::io::AsyncRead`] impl
+pub struct AsyncReadStream<T> {
+    state: StreamState,
+    buffer: Vec<u8>,
+    reader: T,
+}
+
+impl<T> AsyncReadStream<T> {
+    /// Create a [`AsyncReadStream`]. In order to use the [`HostInputStream`] impl
+    /// provided by this struct, the argument must impl [`tokio::io::AsyncRead`].
+    pub fn new(reader: T) -> Self {
+        AsyncReadStream {
+            state: StreamState::Open,
+            buffer: Vec::new(),
+            reader,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<T: tokio::io::AsyncRead + Send + Sync + Unpin + 'static> HostInputStream
+    for AsyncReadStream<T>
+{
+    fn read(&mut self, mut dest: &mut [u8]) -> Result<(u64, StreamState), Error> {
+        use std::io::Write;
+        let l = dest.write(&self.buffer)?;
+
+        self.buffer.drain(..l);
+        if !self.buffer.is_empty() {
+            return Ok((l as u64, StreamState::Open));
+        }
+
+        if self.state.is_closed() {
+            return Ok((l as u64, StreamState::Closed));
+        }
+
+        let dest = &mut dest[l..];
+        let rest = if !dest.is_empty() {
+            let mut readbuf = tokio::io::ReadBuf::new(dest);
+
+            let noop_waker = noop_waker();
+            let mut cx: Context<'_> = Context::from_waker(&noop_waker);
+            // Make a synchronous, non-blocking call attempt to read. We are not
+            // going to poll this more than once, so the noop waker is appropriate.
+            match Pin::new(&mut self.reader).poll_read(&mut cx, &mut readbuf) {
+                Poll::Pending => {}             // Nothing was read
+                Poll::Ready(result) => result?, // Maybe an error occured
+            };
+            let bytes_read = readbuf.filled().len();
+
+            if bytes_read == 0 {
+                self.state = StreamState::Closed;
+            }
+            bytes_read
+        } else {
+            0
+        };
+
+        Ok(((l + rest) as u64, self.state))
+    }
+
+    async fn ready(&mut self) -> Result<(), Error> {
+        if self.state.is_closed() {
+            return Ok(());
+        }
+
+        let mut bytes = core::mem::take(&mut self.buffer);
```

Comment:
This is a bit tricky because the futures returned from `ready` need to all be "cancel safe", so if there's pending data in this buffer then it'll get lost if the `await` point below is cancelled.